### PR TITLE
Add a pin name generator and supporting header files

### PR DESCRIPTION
**Pull request by bremoran**
The pin name generator is invoked by the compilation of the empty pins.c.  Since pins.c depends on ```../generated/pins.array```, that invokes the execution of pins.py, which generates pins.c.

This adds support for yotta-config-based pin definitions.  This is just a prototype, and the pin definition generator probably needs some work.

### CODE REVIEW

**Comment by bremoran** [comment]
This PR depends on:
https://github.com/ARMmbed/target-frdm-k64f-gcc/pull/9
https://github.com/ARMmbed/mbed-hal-k64f/pull/15

**Comment by bremoran** [comment]
This PR will break on STM32 and k64f-armcc until pin definitions are added to those targets and an equivalent of https://github.com/ARMmbed/mbed-hal-k64f/pull/15 is added to STM32

File: source/pins.cmake [hunk 1]
```diff
@@ -0,0 +1,18 @@
+set(pinScript  "${CMAKE_CURRENT_LIST_DIR}/pins.py")
+set(pinFileDir "${CMAKE_CURRENT_LIST_DIR}/../generated")
```
Comment (bogdanm) [inline source/pins.cmake]: Not a real issue, but I'd put `generated/` under `build/`, to make it more clear that the directory contains automatically generated files that are not supposed to be edited. We already have `yotta_config.h` under `build/`.

Comment (AlessandroA) [inline source/pins.cmake]: +1 for moving to `build/`

**Comment by bogdanm** [comment]
+1

**Comment by bremoran** [comment]
There are some more linked PRs coming!

File: source/pins.cmake (hunk 1 above)
Comment (bremoran) [inline source/pins.cmake]: Initially, there were some problems with this, but I'll see if I can resolve them.

**Comment by bremoran** [comment]
@marcuschangarm @rgrover This will affect your targets soon too.

**Comment by marcuschangarm** [comment]
Do you have an example file showing the actual pin configuration?

**Comment by rgrover** [comment]
@bremoran what will be necessary to comply with this?

**Comment by bremoran** [comment]
@rgrover you'll need to make changes equivalent to:
https://github.com/ARMmbed/target-frdm-k64f-gcc/pull/9
https://github.com/ARMmbed/mbed-hal-k64f/pull/15

**Comment by rgrover** [comment]
@bremoran: why are 'pins' defined within 'mbed-os'? To me these are unrelated concepts.

**Comment by bremoran** [comment]
@rgrover Where would you rather see them defined?  They're notionally dependent on infrastructure provided by mbed-hal (e.g. pinmap.h).

**Comment by rgrover** [comment]
It would make more sense to have them in mbed-hal.

**Comment by bremoran** [comment]
I can see an argument for "hal," but I think instead, we should create a configuration section called something like "hw-config".  This could also include information like which hardware exists in the MCU.

**Comment by rgrover** [comment]
yes, that would be much better. thanks.

**Comment by bremoran** [comment]
A quick update: I haven't got a good way of putting the generated file in ```./build```.  I can generate the file there, but determining how to include it is not so simple.  ```YOTTA_GLOBAL_INCLUDE_DIRS``` does not work properly because it is CMake order-dependent in trees with circular dependencies.

The interim solution to this problem appears to be modifying mbed-gcc and mbed-armcc and adding
```CMake
include_directories("${CMAKE_BINARY_DIR}/generated/include")
```

**Comment by bremoran** [comment]
Target support:
* frdm-k64f-gcc: TESTED
* frdm-k64f-armcc: Not tested, but equivalent to frdm-k64f-gcc
* STM32F429I-DISCO: NOT PORTED
* STM32F439 targets: TESTED

**Comment by 0xc0170** [comment]
@bremoran I am reviewing this addition (and references) and I would like to test this locally ,before releasing new version of hal.

**Comment by 0xc0170** [comment]
@bremoran 

I linked mbed-hal-k64f, and this to mbed-drivers, it fails with many errors related to changes done with config - undefined PinMode, GPIO SHIFT and others, an example:
```
/mbed-drivers/yotta_modules/mbed-hal-k64f/mbed-hal/chip_pins.array:17:15: error: 'GPIO_PORT_SHIFT' was not declared in this scope
 PTA0  = (0 << GPIO_PORT_SHIFT | 0 ),
```

Got the latest k64f target, yotta from master, linked mbed-hal (this PR) and mbed-hal-k64f from master. Am I missing something more?

**Comment by miklis** [comment]
Build finished. No test results found.

**Comment by 0xc0170** [comment]
LGTM. Tested locally, works for me. 

I'll merge this today and release a new mbed-hal version.

File: mbed-hal/PinNames.h [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef MBED_PINNAMES_H
+#define MBED_PINNAMES_H
+
+#include "cmsis.h"
+
+#include "mbed-hal/PinDefs.h"
```
Comment (0xc0170) [inline mbed-hal/PinNames.h]: I see below #ifdef YOTTA_CFG_HARDWARE_PINS , if the intention is to keep compatibility, this include should be probably also there? 

Plus this is PinNames header file (guards there), so including PinName.h below does not have any affect?

**Comment by miklis** [comment]
Build finished. No test results found.

**Comment by miklis** [comment]
Build finished. No test results found.

**Comment by 0xc0170** [comment]
Looking good now It builds for multiple targets.

**Comment by 0xc0170** [comment]
This fails for legacy (no config, PinNames defined in the target HAL) targets (freescale/st are working) , where we have now two PinNames.h, the right one does not get in (might be -I order), I recall I did test this (see the message above, when it did not fail :weary: ). Anyway,

For new minor version if this repo we could do
- remove target dependencies which are not supported (nordic, silabs), then we get device.h header not found error, not ideal
- hold this minor release until we find a fix for this (Please someone try to reproduce the problem I am seeing)
- go ahead , release new minor version, we find a fix or just deprecate the old PinNames

**Comment by bremoran** [comment]
The only plausible answer I see to this is to temporarily rename ```mbed-hal/PinNames.h`` and add a redirect PinNames.h to the ST & Freescale HAL.

**Comment by autopulated** [comment]
```C
#ifndef YOTTA_CFG_HARDWARE_PINS
  #ifndef YOTTA_CFG_MBEDOS_LEGACY_PINNAMES_HEADER
    #error "you need to either define the mbedos.legacy_pinnames_header config property, or define pin-names in config"
  #endif
  #include YOTTA_CFG_MBEDOS_LEGACY_PINNAMES_HEADER
#endif // ndef YOTTA_CFG_HARDWARE_PINS
```

things to test about this: forward slashes in YOTTA_CFG_MBEDOS_LEGACY_PINNAMES_HEADER definition work correctly cross-platform (they should do),

**Comment by bremoran** [comment]
Ideas have been brought: https://github.com/ARMmbed/mbed-hal/pull/24

**Pull request merged**